src/modalities/optimizers/optimizer_factory.py

```diff
@@ -95,20 +109,42 @@ def _assert_existence_of_weight_decay_groups_excluded(model: FSDP, weight_decay_
 
 
 def _create_optimizer_groups(
-    model: FSDP, weight_decay: float, weight_decay_groups_excluded: list[str]
+    model: FSDPX, weight_decay: float, weight_decay_groups_excluded: list[str]
 ) -> tuple[OptimizerGroups, list[str]]:
     """
     create optimizer groups of parameters with different weight decays that are to be used in Adam or AdamW
     """
-    nn_model: NNModel = model.module
-    weight_decay_groups = nn_model.weight_decay_groups
-    params = {name: parameter for name, parameter in model.named_parameters() if parameter.requires_grad}
+    # FSDP 1
```

Comment:
Re-use logic:

if isinstance(model, (FSDP1, FSDP2)):
    nn_model = model.module if isinstance(model, FSDP1) else model
    weight_decay_groups = nn_model.weight_decay_groups

    def is_valid_param(name, param):
        if not param.requires_grad:
            return False
        if isinstance(model, FSDP2) and isinstance(param, DTensor):
            return param.to_local().numel() > 0
        return True

    params = {
        name: param
        for name, param in model.named_parameters()
        if is_valid_param(name, param)
    }

else:
    raise OptimizerError(
        f"model {type(model)} is not an instance of FSDP1 or FSDP2. Please use the correct model type."
    )

Reply:
In my opinion the code duplication is very limited and with the introduction of the checks within `is_valid_params` it makes it rather difficult to read. I think having the if condition based on FSDP1 and FSDP2 is fine, since we have different filtering routines.